sros/docker/launch.py

```diff
@@ -773,10 +799,21 @@ def __init__(self, hostname, username, password, mode, variant_name, conn_mode):
                     major_release,
                     variant,
                     conn_mode,
-                ),
-                SROS_lc(variant, conn_mode, variant["max_nics"]),
+                )
             ]
 
+            start_eth = 1
+            for i, lc in enumerate(variant["lcs"]):
+                # Priority is to use max_nics from each lc definition
+                max_nics = lc.get('max_nics', None)
+                if not max_nics:
+                    max_nics = variant['max_nics']
+
+                self.vms.append(
+                    SROS_lc(lc, conn_mode, max_nics, slot=2 + i, nic_eth_start=start_eth)
```

Comment:
I think this needs to be changed
I propose the following

in the `lc` block users provide the slot explicitly (like they would do in a timos line). This slot number is parsed by the code in the `parse_custom_variant` function and the slot information is saved.

The user-provided slot is then used when we call `SROS_lc` function. This will allow users to provide custom slot numbers for their linecards

Reply:
I got your point.
I need to check first. if there any relation between the number of LC VM instances and slot number detected in SROS "show card" command VIM. If TIMOS Line can represent slot number that is detected in VIM, then it can be implemented.

Reply:
This part is already implemented on my latest commit. 
Need your help to review again.

Reply:
I am not sure that we need all those extra functions.
Maybe we can just parse out the slot number in [this function](https://github.com/gusman/vrnetlab/blob/1e7bbee6f086e1f09a3176340ab75011074c6021/sros/docker/launch.py#L268)?

We only need this slot number to pass it in [this function](https://github.com/gusman/vrnetlab/blob/1e7bbee6f086e1f09a3176340ab75011074c6021/sros/docker/launch.py#L886), and nothing else.
In this call it should be just `slot=slot` since the slot number is provided by a user. What do you think?

Reply:
Yes, we can get slot information from that, and simpler.
I did not notice about that, let me re-implement again and test.

On the other hand, in my opinion the sorting procedure is still needed, 
to ensure LC VM instantiation is still in slot sequence, to make sure NICs sequence correct.

Reply:
I am not sure if the sorting is needed @gusman 

Here is the logic I had in mind:

suppose you defined a custom type that goes like this

```
cp: <something> ___
lc: slot=2 max_nic=4 ___
lc: slot=4 max_nic=6
```

when you parse the slot information out, you deploy your VM.
VM_LC1 is the linecard VM that was defined first, its `start_eth_idx = 1`, so the VM will start up with tap1 through tap4 interfaces (if those links are defined in clab file).

VM_LC2 will start with `start_eth_idx=5` and the tap interfaces will be `tap5` through `tap10`.
So regardless which slot the user sets, it is only important to follow the number of nics that each of those LC VMs may have to identify indexes of the corresponding `ethX` and `tapX` interfaces

Reply:
The concern that I have is with the custom type with following sequence:
```
cp: <something> ___
lc: slot=4 max_nic=4 ___
lc: slot=2 max_nic=6
```
Then user can define eth1 -> eth6 for card in slot 2, and eth7 -> eth11 for card in slot 4.
In that case, it needs to ensure that start_eth_idx = 1 belong to first port of card in slot 2.

My idea is to sort the lc definition based on the slot value then start_eth_idx value can follow up.
Using the idea that you share before, I believe sorting logic can be simpler that in my last commit.

Let me know your thought.

Reply:
okay, I assumed that users should take this on themselves (sorting order), but if you coded it up, than it is even better.




> Also, can we exclude LC lines without slot token from LCS list, since VM instantiation will fail if slot information misising?

which list are you referring to?

Reply:
> okay, I assumed that users should take this on themselves (sorting order), but if you coded it up, than it is even better.
> 
> > Also, can we exclude LC lines without slot token from LCS list, since VM instantiation will fail if slot information misising?
> 
> which list are you referring to?
The `variant["lcs"]` that contains list of line cards definition.
My Idea, if a line card definition does not have slot information, then do not append it into the `variant["lcs"]`.

Reply:
we should use the slot number in the `lcs` for the embedded variants
we can use slot=1 for all of them